### PR TITLE
Improve testability and add tests to the way that blocks of coefficients are encoded

### PR DESCRIPTION
**Pull request by mcroomp**
add read_coefficients and write_coefficients that encapsulate the read/write process and can be testing individually without pulling in the world.

This also allows for testing with hardcoded known good values to make it easier to catch any subtle changes that could cause backwards compat issues.

### CODE REVIEW

**Comment by Melirius** [comment]
LGTM

File: src/structs/lepton_decoder.rs [hunk 1]
```diff
@@ -289,30 +289,49 @@ fn parse_token<R: Read, const ALL_PRESENT: bool>(
 ) -> Result<()> {
     debug_assert!(pt.is_all_present() == ALL_PRESENT);
 
+    let neighbors =
+        context.get_neighbor_data::<ALL_PRESENT>(image_data, context, num_non_zeros, pt);
```
Comment (Melirius) [inline src/structs/lepton_decoder.rs:293]: What about renaming `num_non_zeros`? It is not the only and not even the main component of `NeighborSummary`.

**Comment by mcroomp** [comment]
Yeah that’s a dumb name agree

File: src/structs/lepton_decoder.rs (hunk 1 above)
Comment (Melirius) [reply src/structs/lepton_decoder.rs:293]: Also `context` is already there, should not send it inside the function again.

File: src/structs/probability_tables.rs [hunk 2]
```diff
@@ -92,17 +91,16 @@ impl ProbabilityTables {
 
     pub fn calc_non_zero_counts_context_7x7<const ALL_PRESENT: bool>(
```
Comment (Melirius) [inline src/structs/probability_tables.rs:92]: I think: what if rename this to `calc_non_zero_counts_predictor_7x7` or smth like that and return `num_non_zeros_above << 1`,  `num_non_zeros_left << 1` or `num_non_zeros_above + num_non_zeros_leftnum_non_zeros_above + num_non_zeros_left. Then by extending array `NON_ZERO_TO_BIN` to 100 bytes we can reduce number of operations per coefficient.

**Comment by Melirius** [comment]
Nice!

File: src/structs/model.rs [hunk 3]
```diff
@@ -46,6 +46,99 @@ pub struct Model {
     counts_dc: [CountsDC; NUMERIC_LENGTH_MAX],
 }
 
+impl Model {
+    /// walks through the model and applies the walker function to each branch
```
Comment (Melirius) [inline src/structs/model.rs:50]:
```suggestion
    /// Walks through the model and applies the walker function to each branch
```

File: src/structs/probability_tables.rs (hunk 2 above)
Comment (mcroomp) [reply src/structs/probability_tables.rs:92]: I'll separate this in another change... better to separate refactoring from functional changes

File: src/structs/lepton_decoder.rs [hunk 4]
```diff
@@ -366,6 +382,8 @@ pub fn read_coefficients<const ALL_PRESENT: bool, R: Read>(
             output.set_coefficient(coord as usize, coef);
         }
     }
+
```
Comment (Melirius) [inline src/structs/lepton_decoder.rs:385]: Here we can add another check for zero `num_non_zeros_left_7x7`

File: src/structs/lepton_decoder.rs [hunk 5]
```diff
@@ -282,77 +282,120 @@ fn parse_token<R: Read, const ALL_PRESENT: bool>(
...
+    features: &EnabledFeatures,
+) -> Result<(AlignedBlock, NeighborSummary)> {
     let model_per_color = model.get_per_color(pt);
 
+    // First we read the 49 inner coefficients
+
+    // figure out how many of these are non-zero, which is used both
+    // to terminate the loop early and as a predictor for the model
+    let predicted_num_non_zeros_7x7 =
+        pt.calc_non_zero_counts_context_7x7::<ALL_PRESENT>(neighbor_data);
```
Comment (Melirius) [inline src/structs/lepton_decoder.rs:325]: Hmm, the names are slightly misleading, as the number returned is from 0 to 25 - definitely not number of zeros, but just its predictor. What about `predicted_num_non_zeros_bin_7x7` and return bin? With corresponding changes in `read_non_zero_7x7_count` and in writer?

Comment (mcroomp) [reply src/structs/lepton_decoder.rs]: made it num_non_zeros_7x7_context_bin

Comment (mcroomp) [reply src/structs/lepton_decoder.rs]: Feel free to approve the review if you are ok with the changes

**Review by Melirius — APPROVED** [review]
Perfect!

**Review by gbrovman — APPROVED** [review]

**Pull request merged**